README
what is this?
=============

rotatedaemon is a C++ conversion (and enhancement) of a script called 'rotate.py' for Maemo (original author unknown).

Basically, it rotates applications - and not just landscape or portrait, reverse landscape/portrait are supported too.

It also supports a granular blacklist feature (to avoid rotating unstable applications), unlike rotate.py.

I was going to do a fair bit more with this, but really, there aren't many applications that work well in portrait mode.

contact
=======

All contact to do with this code should go to:
 robin.burchell@collabora.co.uk

Patches welcome.


debian/changelog
rotatedaemon (0.1-4) unstable; urgency=low

  * Add Upstart job to start rotatedaemon on hildon-desktop start.

 -- Robin Burchell <viroteck@viroteck.net>  Fri, 18 Jun 2010 19:19:10 +0100

rotatedaemon (0.1-3) unstable; urgency=low

  * Don't ask.

 -- Robin Burchell <viroteck@viroteck.net>  Fri, 18 Jun 2010 19:01:11 +0100

rotatedaemon (0.1-2) unstable; urgency=low

  * Fixes to make it even remotely functional.
    Apparently this is a very old version of my code. :(

 -- Robin Burchell <viroteck@viroteck.net>  Fri, 18 Jun 2010 16:00:31 +0100

rotatedaemon (0.1-1) unstable; urgency=low

  * Fix project file to install binary correct.

 -- Robin Burchell <viroteck@viroteck.net>  Fri, 18 Jun 2010 15:24:22 +0100

rotatedaemon (0.1) unstable; urgency=low

  * Initial release

 -- Robin Burchell <viroteck@viroteck.net>  Fri, 18 Jun 2010 14:56:00 +0100


main.cpp
/*
 * Copyright (C) 2010 Robin Burchell <robin.burchell@collabora.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU Lesser General Public License,
 * version 2.1, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "rotationdaemon.h"

int main(int argc, char **argv)
{
    RotationDaemon a(argc, argv);

    a.exec();
}


rotationdaemon.h
/*
 * Copyright (C) 2010 Robin Burchell <robin.burchell@collabora.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU Lesser General Public License,
 * version 2.1, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 */

#ifndef ROTATIONDAEMON_H
#define ROTATIONDAEMON_H

#include <QApplication>
#include <QFile>
#include <QPoint>
#include <X11/Xatom.h>
#include <X11/Xlib.h>

class RotationDaemon : public QApplication
{
    Q_OBJECT
public:
    explicit RotationDaemon(int argc, char **argv);
private slots:
    void timerEvent();
private:
    QPoint getOrientationInformation();
    void rotate(int angle);

    Window activeWindowId();
    bool checkApplicationBlacklistToForceLandscape();

    QFile m_accelerometer;
    Atom m_NET_ACTIVE_WINDOW;
};

#endif // ROTATIONDAEMON_H


rotationdaemon.cpp
/*
 * Copyright (C) 2010 Robin Burchell <robin.burchell@collabora.co.uk>
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU Lesser General Public License,
 * version 2.1, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <QApplication>
#include <QX11Info>
#include <QDebug>
#include <QTimer>
#include <QFile>

#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xatom.h>
#include <X11/extensions/Xrandr.h>

#include "rotationdaemon.h"

RotationDaemon::RotationDaemon(int argc, char **argv) :
    QApplication(argc, argv),
    m_accelerometer(QLatin1String("/sys/class/i2c-adapter/i2c-3/3-001d/coord"))
{
    if (!m_accelerometer.open(QFile::ReadOnly)) {
        qFatal("Cannot open accelerometer FD");
    }

    m_NET_ACTIVE_WINDOW = XInternAtom(QX11Info::display(), "_NET_ACTIVE_WINDOW", False);
    if (m_NET_ACTIVE_WINDOW == None) {
        qFatal("_NET_ACTIVE_WINDOW is apparantly None?");
    }

    QTimer *t = new QTimer(this);
    connect(t, SIGNAL(timeout()), SLOT(timerEvent()));
    t->start(1000);
}

void RotationDaemon::rotate(int angle)
{
    // XXX: technically this is leaked
    static XRRScreenConfiguration *config = XRRGetScreenInfo(QX11Info::display(), DefaultRootWindow(QX11Info::display()));

    ushort rotation;
    ushort size;
    size = XRRConfigCurrentConfiguration(config, &rotation);

    short rate;
    rate = XRRConfigCurrentRate(config);

    unsigned long timestamp, current_time;
    timestamp = XRRTimes(QX11Info::display(), XDefaultScreen(QX11Info::display()), &current_time);

    qDebug() << "RotationDaemon::rotate(" << angle << ")";
    XRRSetScreenConfigAndRate(QX11Info::display(), config, DefaultRootWindow(QX11Info::display()), size, angle, rate, timestamp);
}

void RotationDaemon::timerEvent()
{
    if (checkApplicationBlacklistToForceLandscape()) {
        // TODO: if we're not already landscape, make us so.
        return;
    }
    
    QPoint p = getOrientationInformation();
    qDebug() << p;

    if (p.y() < -500) {
        rotate(1);
    } else if (p.y() > 500) {
        rotate(4);
    } else if (p.x() < -500) {
        rotate(2);
    } else if (p.x() > 500) {
        rotate(8);
    }
}

QPoint RotationDaemon::getOrientationInformation()
{
    QPoint p;

    // rewind
    m_accelerometer.seek(0);

    QByteArray bytes = m_accelerometer.readAll();
    QList<QByteArray> tokens = bytes.split(' ');

    p.setX(tokens[0].toInt());
    p.setY(tokens[1].toInt());

    return p;
}

Window RotationDaemon::activeWindowId()
{
    Window active_win = None;
    Atom actual_type;
    int actual_format;
    unsigned long nitems, bytes_after;
    unsigned char *prop_return = NULL;

    if(Success == XGetWindowProperty(QX11Info::display(), DefaultRootWindow(QX11Info::display()), m_NET_ACTIVE_WINDOW, 0L, sizeof(Window),
                                     False, XA_WINDOW, &actual_type,
                                     &actual_format, &nitems, &bytes_after,
                                     &prop_return) && prop_return)
    {
        active_win = *(Window *)prop_return;
        XFree(prop_return);
    }

    return active_win;
}


bool RotationDaemon::checkApplicationBlacklistToForceLandscape()
{
    // Find active window
    Window activeWindow = activeWindowId();

    // Now find what it is called
    XClassHint *xh = XAllocClassHint();
    //Status s = 
    XGetClassHint(QX11Info::display(), activeWindow, xh);

    qDebug() << "Active window is " << xh->res_name << xh->res_class;

    // clean up
    XFree(xh->res_name);
    XFree(xh->res_class);
    XFree(xh);

    return false;
}
